### PR TITLE
[IMP] hr_timesheet,project,sale_timesheet,sale_project,website_sale_slides: generi…

### PR DESCRIPTION
**Pull request by matu-odoo**
Purpose of the PR is to do the generic improvements for project.

So in this PR did the following changes:
- add a space between 'access to:' and the name of the course in SOL nameget.
- change the second part of the warning in project form view.
- hide the 'remaining hours' value if the nb of 'initially planned hours' is
  equal to 0 in task list view.
- hide the 'initially planned hours' and 'hours spent' values if they are equal
  to 0 in task list view.
- hide the 'milestone' and 'assignees' fields if they are false in task calendar
  popover.
- hide the 'remaining hours on SO' value if there is no SOL set on the task in
  task list view.
- use the progressbar widget for the progress field in project update list view.
- add the fields effective hours/days, timesheets progress, milestones progress,
  next milestone in project list view.

task - 3178057

### CODE REVIEW

**Review by auon-odoo — CHANGES_REQUESTED** [review]
thank you :fire:

**Comment by kcv-odoo** [comment]
Done! Thanks :)

**Review by auon-odoo — APPROVED** [review]
lgtm :fire:

File: addons/hr_timesheet/models/project_project.py [hunk 1]
```diff
@@ -38,6 +40,11 @@ def _compute_allow_timesheets(self):
         without_account = self.filtered(lambda t: not t.analytic_account_id and t._origin)
         without_account.update({'allow_timesheets': False})
 
+    @api.depends('allocated_hours', 'remaining_hours')
+    def _compute_remaining_hours_percentage(self):
+        for project in self:
+            project.remaining_hours_percentage = project.remaining_hours / project.allocated_hours if project.allocated_hours else 0.0
```
Comment (auon-odoo) [inline addons/hr_timesheet/models/project_project.py:46]:
```suggestion
            project.remaining_hours_percentage = project.allocated_hours and project.remaining_hours / project.allocated_hours
```

File: addons/project/models/project_project.py [hunk 2]
```diff
@@ -208,11 +208,26 @@ def _read_group_stage_ids(self, stages, domain, order):
...
+    def _compute_next_milestone_id(self):
+        for project in self:
+            next_milestones = project.milestone_ids.filtered(lambda m: not m.is_reached)
+            project.next_milestone_id = next_milestones and next_milestones[0].id
+
+    @api.depends('next_milestone_id.is_reached', 'next_milestone_id.deadline')
+    def _compute_is_next_milesotone_unreached_and_late(self):
+        for project in self:
+            next_milestone = project.next_milestone_id
+            project.is_next_milesotone_unreached_and_late = not next_milestone.is_reached and next_milestone.deadline and next_milestone.deadline < fields.Date.today()
```
Comment (auon-odoo) [inline addons/project/models/project_project.py:229]:
```suggestion
        today = fields.Date.today()
        for project in self:
            next_milestone = project.next_milestone_id
            project.is_next_milesotone_unreached_and_late =\
                not next_milestone.is_reached and\
                next_milestone.deadline and\
                next_milestone.deadline < today
```
- Better to break the line imo
- Get `fields.Date.today()` outside the loop

File: addons/project/models/project_project.py [hunk 3]
```diff
@@ -270,6 +285,7 @@ def _compute_milestone_reached_count(self):
         mapped_count = {project.id: count for project, count in read_group}
         for project in self:
             project.milestone_count_reached = mapped_count.get(project.id, 0)
+            project.milestone_progress = project.milestone_count_reached * 100 // project.milestone_count if project.milestone_count > 0 else 0
```
Comment (auon-odoo) [inline addons/project/models/project_project.py:288]:
```suggestion
            project.milestone_progress = project.milestone_count and project.milestone_count_reached * 100 // project.milestone_count
```

File: addons/project/models/project_project.py [hunk 4]
```diff
@@ -208,11 +208,26 @@ def _read_group_stage_ids(self, stages, domain, order):
...
 
     _sql_constraints = [
         ('project_date_greater', 'check(date >= date_start)', "The project's start date must be before its end date.")
     ]
 
+    @api.depends('milestone_ids', 'milestone_ids.is_reached', 'milestone_ids.deadline')
+    def _compute_next_milestone_id(self):
+        for project in self:
+            next_milestones = project.milestone_ids.filtered(lambda m: not m.is_reached)
+            project.next_milestone_id = next_milestones and next_milestones[0].id
```
Comment (auon-odoo) [inline addons/project/models/project_project.py:223]:
```suggestion
            project.next_milestone_id = project.milestone_ids.filtered(
                lambda m: not m.is_reached
            )[:1].id
```
But is sounds a little too costly to filter (1) in a loop, (2) if we only take the first result.
Do you think of another solution?

File: addons/project/models/project_project.py [hunk 5]
```diff
@@ -208,11 +208,26 @@ def _read_group_stage_ids(self, stages, domain, order):
     milestone_count = fields.Integer(compute='_compute_milestone_count', groups='project.group_project_milestone')
     milestone_count_reached = fields.Integer(compute='_compute_milestone_reached_count', groups='project.group_project_milestone')
     is_milestone_exceeded = fields.Boolean(compute="_compute_is_milestone_exceeded", search='_search_is_milestone_exceeded')
+    milestone_progress = fields.Integer("Milestones Reached", compute='_compute_milestone_reached_count')
+    next_milestone_id = fields.Many2one('project.milestone', compute='_compute_next_milestone_id', store=True)
```
Comment (auon-odoo) [inline addons/project/models/project_project.py:212]: Why store it?

**Review by auon-odoo — CHANGES_REQUESTED** [review]
Thanks for your work!

File: addons/project/models/project_project.py (hunk 5 above)
Comment (roku-odoo) [reply addons/project/models/project_project.py:212]: Opps! we need to make it store because we are accessing `Field.AnotherField` in depend and we can not access field via `dot` on non store field.
(P.S. Rationale from red runbot :see_no_evil: )

Comment (auon-odoo) [reply addons/project/models/project_project.py:212]: Mh ok, I didn't know that!

File: addons/project/models/project_project.py (hunk 4 above)
Comment (auon-odoo) [reply addons/project/models/project_project.py]: What about this? Isn't a _read_group better, performance-wise?

Comment (roku-odoo) [reply addons/project/models/project_project.py]: Changes Done!

File: addons/project/models/project_project.py [hunk 6]
```diff
@@ -208,11 +208,34 @@ def _read_group_stage_ids(self, stages, domain, order):
...
 
+    @api.depends('milestone_ids', 'milestone_ids.is_reached', 'milestone_ids.deadline')
+    def _compute_next_milestone_id(self):
+        projects_read_group = {project.id: milestone_list for project, milestone_list in self.env['project.milestone']._read_group(
+            [('project_id', 'in', self.ids), ('is_reached', '=', False)],
+            ['project_id'],
+            ['id:array_agg'],
+        )}
+        for project in self:
+            project.next_milestone_id = projects_read_group.get(project.id, False) and projects_read_group[project.id][0]
```
Comment (auon-odoo) [inline addons/project/models/project_project.py:227]:
```suggestion
        milestone_ids_per_project_id = {
            project.id: milestone_ids
            for project, milestone_ids in self.env['project.milestone']._read_group(
                [('project_id', 'in', self.ids), ('is_reached', '=', False)],
                ['project_id'],
                ['id:array_agg'],
            )}
        }
        for project in self:
            project.next_milestone_id = project.id in milestone_ids_per_project_id and milestone_ids_per_project_id[project.id][0]
```
Glad you found a solution :smiley: Would just improve it a little bit:
- Name the dict `<key>_per_<value>`
- Some indents
- When you only want to check if in dict, use `key in dict` rather than `dict.get(key)`

**Comment by auon-odoo** [comment]
task id is wrong :/

**Review by auon-odoo — COMMENTED** [review]
Thanks! Green the task when you're done with this last change

File: addons/project/models/project_project.py (hunk 6 above)
Comment (roku-odoo) [reply addons/project/models/project_project.py]: Done!

**Review by auon-odoo — APPROVED** [review]
Lgtm, thanks! :slightly_smiling_face:

**Comment by auon-odoo** [comment]
But the task id is still wrong

**Comment by xavierbol** [comment]
It seems you have some conflicts, can you resolve it when you have time? 

Thanks in advance! 🙂

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I left some comments. 🙂 

I will push a commit and rebase your branch since there is a conflict. 🙂 

If you agree with that then you can squash the commit with yours. 🙂

File: addons/project/models/project_project.py [hunk 7]
```diff
@@ -208,11 +209,28 @@ def _read_group_stage_ids(self, stages, domain, order):
...
+        milestone_ids_per_project_id = {
+            project.id: milestone_ids
+            for project, milestone_ids in self.env['project.milestone']._read_group(
+                [('project_id', 'in', self.ids), ('is_reached', '=', False)],
+                ['project_id'],
+                ['id:array_agg'],
+            )
+        }
+        for project in self:
+            project.next_milestone_id = project.id in milestone_ids_per_project_id and milestone_ids_per_project_id[project.id][0]
```
[the remark this answers is not in the document]
Comment (ppr-odoo) [reply addons/project/models/project_project.py]: > 


A traceback will occur when a project does not have a milestone so I fixed it in the my last commit.

**Comment by ppr-odoo** [comment]
> Thanks for your work, I left some comments. 🙂
> 
> I will push a commit and rebase your branch since there is a conflict. 🙂
> 
> If you agree with that then you can squash the commit with yours. 🙂


I have made some changes in the last commit. Please lookat it. Then I'll squash it into a single commit.

File: addons/project/models/project_project.py [hunk 8]
```diff
@@ -242,10 +242,10 @@ def _compute_next_milestone_id(self):
...
         }
         for project in self:
-            next_milestone = milestone_ids_per_project_id[project.id][:1]
-            project.next_milestone_id = next_milestone
-            project.can_mark_milestone_as_done = next_milestone.can_mark_milestone_as_done
-            project.is_milestone_deadline_exceeded = next_milestone.is_milestone_deadline_exceeded
+            milestone = milestone_ids_per_project_id.get(project.id)
+            project.next_milestone_id = milestone and milestone[:1]
+            project.can_mark_milestone_as_done =  milestone and milestone[:1].is_deadline_exceeded
+            project.is_milestone_deadline_exceeded =  milestone and milestone[:1].is_deadline_exceeded
```
Comment (xavierbol) [inline addons/project/models/project_project.py:248]: Alternatively, we could do something like that: 

```suggestion
            milestone = milestone_ids_per_project_id.get(project.id, self.env['project.milestone'])
            project.next_milestone_id = milestone[:1]
            project.can_mark_milestone_as_done = milestone.can_be_marked_as_done
            project.is_milestone_deadline_exceeded = milestone.is_deadline_exceeded
```

File: addons/project/views/project_project_views.xml [hunk 9]
```diff
@@ -238,12 +238,12 @@
                     <field name="date_start" string="Planned Date" widget="daterange" options="{'end_date_field': 'date'}" required="False" optional="hide"/>
                     <field name="date" column_invisible="True" />
                     <field name="milestone_progress" widget="progressbar"
-                        invisible="milestone_progress == 0 or not allow_milestones"
+                        column_invisible="milestone_progress == 0"
```
Comment (xavierbol) [inline addons/project/views/project_project_views.xml:241]: why do we no longer use `allow_milestones`?

Why do you use `column_invisible` instead of `invisible`?

File: addons/project/views/project_project_views.xml [hunk 10]
```diff
@@ -238,12 +238,12 @@
...
                         optional="hide"
                     />
+                    <field name="is_milestone_deadline_exceeded" invisible="1"/>
                     <field name="next_milestone_id"
-                        decoration-danger="is_deadline_exceeded" decoration-success="can_be_marked_as_done"
-                        invisible="not allow_milestones"
```
Comment (xavierbol) [inline addons/project/views/project_project_views.xml]: Why did you remove the invisible?

**Review by xavierbol — COMMENTED** [review]
Thanks for the update. 🙂 

I left some comments. 🙂

File: addons/project/models/project_project.py (hunk 8 above)
Comment (ppr-odoo) [reply addons/project/models/project_project.py]: > Alternatively, we could do something like that:

Done

File: addons/project/views/project_project_views.xml (hunk 10 above)
Comment (ppr-odoo) [reply addons/project/views/project_project_views.xml]: > Why did you remove the invisible?

Similar to the above comment.

**Review by xavierbol — COMMENTED** [review]
LGTM, thanks you can squash the commits. 🙂 

I just left one comment to improve my previous suggestion. 🙂

**Comment by ppr-odoo** [comment]
> LGTM, thanks you can squash the commits. 🙂
> 
> I just left one comment to improve my previous suggestion. 🙂

done.thanks

**Review by xavierbol — APPROVED** [review]
LGTM, thanks again for your work. 🙂 

robodoo r+
